Add support for Windows Dark Mode

davidhewitt opened the pull request:
I had a go at adding simple support for Windows 10's "Dark Mode". The moment the title bar will set to light / dark automatically according to system theme. I did not add an option for users of `winit` to disable this behaviour, as I felt it's likely desirable for most use cases. I can change that if you prefer otherwise!

Please advise on any documentation you'd like me to update.

The "window" example when Dark Mode is enabled:

![image](https://user-images.githubusercontent.com/1939362/66387870-cd5c1c00-e9bc-11e9-942c-9f22ddb780d3.png)

- [x] Tested on all platforms changed
- [x] Compilation warnings were addressed
- [x] `cargo fmt` has been run on this branch
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [ ] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [ ] Created or updated an example program if it would help users understand this functionality
- [ ] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

Closes #1193

davidhewitt commented:
(I rebased onto master to fix the merge conflict)

Osspial commented:
For the most part, this seems good. Could you add a function to `WindowExtWindows` to allow users to retrieve whether dark mode is active or not? We should let users change their theme based on the OS option if they can.

davidhewitt commented:
Thanks for the review! I've added a field to `WindowState` to track the current state of dark mode.

Regarding the event, I was also wondering about such. I know MacOS also has a dark mode though know nothing about its integration in APIs. So the message might be relevant on at least two platforms.

chrisduerr commented:
> I know MacOS also has a dark mode though know nothing about its integration in APIs

Dark mode on macOS is enabled by setting a property in the info.plist file. I'm not sure how you query for it as an application, but to support it there's no support necessary from winit (Alacritty already sets this property).

davidhewitt commented:
A quick google for macOS suggests we might be able to recieve a system broadcast for it on that platform too: https://stackoverflow.com/questions/39048894/how-to-detect-switch-between-macos-default-dark-mode-using-swift-3

So it might be appropriate to add a `WindowEvent` called `SystemThemeChanged` or `DarkModeChanged` - which might never get sent on some platforms (for now) but could be sent on (at least) Windows and macOS.

I don't have a mac so can't do the implementation for that platform. If you agree we should add a new `WindowEvent` I can do the Windows implementation as part of this PR and then open an issue for someone else to follow up with mac support.

Osspial commented:
@davidhewitt I'm fine with taking that approach.

davidhewitt commented:
Great - have now added an event `DarkModeChanged` and confirmed it fires during the `window` example.

davidhewitt commented:
Just a heads up; I recently updated my system to the latest Windows Insiders and it appears that this definition: `const DWMWA_USE_IMMERSIVE_DARK_MODE: DWORD = 19;` now needs to be `20` for the `window` example to run successfully on my system. I've no idea if this is an intended change or a bug or what.

The windows terminal source doesn't appear to have changed from `19` to `20`, but that still works fine, which is puzzling. Might be worth holding off on merging this until I figure out what's going on.

Osspial commented on src/event.rs:
```diff
@@ -204,6 +204,13 @@ pub enum WindowEvent {
+    /// The parameter is true if the window is now in a dark theme, and false if it is not.
+    /// Applications might wish to react to this to change the theme of the content of the window
+    /// when the system changes the window theme.
+    DarkModeChanged(bool),
```
I don't think we should have this be a `bool`, since I can see a there isn't a clear semantic mapping between `true`/`false` and `Dark`/`Light`. Could we have this be `ThemeChanged(Theme)` where `Theme` is defined as follows?

```rust
enum Theme {
    Light,
    Dark,
}
```

davidhewitt commented:
I've updated the to use a `ThemeChanged` event, and created a workaround my problem. The best solution I could come up with is checking the Windows 10 build version and subsequently choosing the appropriate value for `DWMA_USE_IMMERSIVE_DARK_MODE`.

goddessfreya commented:
@Osspial This is, afaik, good for merge?

Osspial commented:
@davidhewitt Sorry it took so long to get back to you on this. I have no idea what's going on with dark mode on your system, since `8ba2055` completely broke this for me. There are two primary issues:

- `WIN10_BUILD_VERSION` is defined as `4026549603` on my system, far outside the range defined in the source file.
- Setting `DWMWA_USE_IMMERSIVE_DARK_MODE` to `20` causes the `DwmSetWindowAttribute` to fail. `19` works fine.

I'm inclined to remove the windows version check and publish this with `DWMWA_USE_IMMERSIVE_DARK_MODE = 19`.

davidhewitt commented:
> I'm inclined to remove the windows version check and publish this with DWMWA_USE_IMMERSIVE_DARK_MODE = 19.

I would happily undo the version check! Strange about my system then. Are you on a stable windows 10 release or an insiders build?

davidhewitt commented:
Rebased on master.

@Osspial I've changed the method used for querying the Windows 10 build version. Does this new method now return a build version within the expected range on your machine?

(I'm currently on Windows 10 insiders build 19018, and `WIN10_BUILD_VERSION` is coming out as `19018`)

If this new method doesn't work on your box, then I'll remove the version checking and hard-code `19` again. Though it might be great if someone else on Windows 10 insiders can report their build number and also whether `19` or `20` is the correct definition of `DWMWA_USE_IMMERSIVE_DARK_MODE` for them.

chelmich commented:
On my machine `WIN10_BUILD_VERSION` is reported as `18362`. This matches what is displayed by `winver`. I can confirm that the dark theme works as intended, including changing the theme while a window is open.

daxpedda commented:
I had access to a machine with windows build 17763 on it, tried this out and it works as intended, `WIN10_BUILD_VERSION` was also reported correctly.

A tiny thing that I noticed: it doesn't change theme when the window isn't focused, but it will switch to the correct one as soon as it's focused again.

davidhewitt commented:
Thanks guys. Anyone else also on Windows 10 insiders builds greater than  `18363`? Looks like if there was an API change (if there is indeed one and it's not my system) it ocurred on a build later than that.

> A tiny thing that I noticed: it doesn't change theme when the window isn't focused, but it will switch to the correct one as soon as it's focused again.

Interesting. Did you happen to notice at what time the `ThemeChanged` event is being fired in the `window` example? Is it when it switches into focus, or when you change the dark theme setting? (On my system the ThemeChanged event fires when I change the system setting, regardless of focus.)

daxpedda commented:
> Did you happen to notice at what time the `ThemeChanged` event is being fired in the `window` example?

It fires as soon as I change the system setting, even though it will only actually switch theme when I focus the window again.

Same behaviour as on your system.

davidhewitt commented:
> It fires as soon as I change the system setting, even though it will only actually switch theme when I focus the window again. 

Cool. So we can probably make it switch theme sooner if we force Windows to redraw the title bar.

Past experiments I've had into trying to do that have never been very successful... anyone know a nice API to force the title bar to redraw? 😄

daxpedda commented on src/platform_impl/windows/event_loop.rs:
```diff
@@ -1895,6 +1896,27 @@ unsafe extern "system" fn public_window_callback<T>(
+                let theme = if is_dark_mode { Dark } else { Light };
+
+                window_state.is_dark_mode = is_dark_mode;
+                subclass_input.send_event(Event::WindowEvent {
```
```suggestion
                mem::drop(window_state);
                subclass_input.send_event(Event::WindowEvent {
```

Before sending an event we should drop `window_state`, trying to do anything on `ThemeChanged` like changing window size will otherwise crash the program.

davidhewitt replied:
Good catch - committed!

daxpedda commented:
> Past experiments I've had into trying to do that have never been very successful... anyone know a nice API to force the title bar to redraw? 😄

I was playing around with this for a bit and as far as I could find, using `SetWindowPos` with `SWP_FRAMECHANGED` should fix this, but it doesn't. The only thing I could find is resizing the window to a new size, any size does it, just not the original one.

So the only workaround I could come up with, is if the window isn't focused, we want to resize the window by one pixel and than back again.

daxpedda commented:
> Given that this focus issue only applies to an out-of-date Windows 10 build

I didn't know this issue is not present in newer builds? I will try on a newer build now and report back.

davidhewitt commented:
> I didn't know this issue is not present in newer builds? I will try on a newer build now and report back.

I'm now doubting myself. Pretty sure on my system (most recent Windows Insiders build) the frame changes whether focussed or not. I'll check later.

daxpedda commented:
You were right, this only affects older versions of windows. I would consider this a bug on windows, not our problem.

davidhewitt commented:
I asked about the support for insiders builds at https://github.com/ysc3839/win32-darkmode/issues/2

From that I learned of a different solution which works for builds `18362` (April 2019) and up, including my latest windows insiders build.

So I've changed the implementation here to that new solution. I've also moved the `lazy_static!` calls around to put the statics closer to where they're used, instead of global.

I've kept the windows 10 build version check in, so that this code will no-op rather than crash on older windows 10 builds.

This is hopefully now good to go!

daxpedda commented on src/platform_impl/windows/dark_mode.rs:
```diff
@@ -0,0 +1,257 @@
+        // We won't try to do anything for windows versions < 17763
+        // (Windows 10 October 2018 update)
+        match *WIN10_BUILD_VERSION {
+            Some(v) => v > 18362,
```
```suggestion
            Some(v) => >= 17763,
```

I think is what you meant?

daxpedda commented:
I tried on `17763` and `18362`, both work with the fix I suggested.

davidhewitt commented:
> I tried on `17763` and `18362`, both work with the fix I suggested.

Thanks! The example I was using only used this method from `18362`, so I was being conservative with putting `18362` in the check. But I forgot to update the comment. Fortunately it turns out `17763` also works 😄 

Out of curiosity, how are you testing so many Win10 versions so quickly? VMs?

daxpedda commented:
> Out of curiosity, how are you testing so many Win10 versions so quickly? VMs?

Hahaha, that would be epic! 😄
My main windows machine is running LTSC, which is currently stuck at `17763`, but my tablet is running `18362`, so we just happen to be fortunate 😁.

daxpedda commented on src/platform_impl/windows/dark_mode.rs:
```diff
@@ -0,0 +1,257 @@
+
+                const UXTHEME_SHOULDAPPSUSEDARKMODE_ORDINAL: WORD = 132;
+
+                let module = libloaderapi::LoadLibraryExW(
```
We are using `LoadLibraryA` in the rest of the library, not sure if this is important to anybody to keep this uniform.

davidhewitt commented:
Thanks; I switched to use `get_function!` and `LoadLibraryA`.

daxpedda commented on src/platform_impl/windows/dark_mode.rs:
```diff
@@ -0,0 +1,222 @@
+            szCSDVersion: [WCHAR; 128],
+        }
+
+        type RtlGetVersion = unsafe extern "system" fn (*mut OSVERSIONINFOW);
```
`RtlGetVersion` has a return type, it's supposed to be `winapi::shared::ntdef::NTSTATUS`.

daxpedda commented on src/platform_impl/windows/dark_mode.rs:
```diff
@@ -0,0 +1,222 @@
+                    szCSDVersion: [0; 128],
+                };
+
+                (rtl_get_version)(&mut vi as _);
```
We should use `winapi::shared::ntdef::NT_SUCCESS` to check if the return value is a success.

daxpedda commented on src/platform_impl/windows/dark_mode.rs:
```diff
@@ -0,0 +1,222 @@
+    if let Some(set_window_composition_attribute) = *SET_WINDOW_COMPOSITION_ATTRIBUTE {
+        unsafe {
+            // SetWindowCompositionAttribute needs a bigbool (i32), not bool.
+            let mut is_dark_mode_bigbool = is_dark_mode as BOOL;
```
```suggestion
            let mut is_dark_mode_bigbool = is_dark_mode.into();
```
Simple widening, doesn't need `as`.

Osspial replied:
I have to thoroughly disagree with this comment, actually. Casting via `as` makes it clear that it's *just* a vanilla cast with nothing fancy going on, which is much less clear with an `into()` call.

daxpedda commented on src/platform_impl/windows/dark_mode.rs:
```diff
@@ -0,0 +1,222 @@
+            let mut data = WINDOWCOMPOSITIONATTRIBDATA {
+                Attrib: WCA_USEDARKMODECOLORS,
+                pvData: &mut is_dark_mode_bigbool as *mut _ as _,
+                cbData: std::mem::size_of_val(&is_dark_mode_bigbool) as _
```
```suggestion
                cbData: std::mem::size_of_val(&is_dark_mode_bigbool).into()
```
The underlying type is `usize`.

davidhewitt replied:
Unfortunately windows wants `u32` for the cbData value, so we need to do a narrowing cast here. Size of a bool is always going to fit safely in a `u32`, so this narrowing `as` isn't dangerous.

davidhewitt commented:
Thanks again, have addressed all those suggestions.

goddessfreya commented on src/event.rs:
```diff
@@ -201,6 +201,12 @@ pub enum WindowEvent {
+    ///
+    /// Applications might wish to react to this to change the theme of the content of the window
+    /// when the system changes the window theme.
+    ThemeChanged(Theme),
```
We should probably mention that it is currently only implemented for windows.

davidhewitt replied:
Thanks, have done so!

VideoPlayerCode commented:
Looking really good now. Great work @davidhewitt

Osspial commented:
Sorry about the delay reviewing this. Once my nit is addressed this should be good to go. Thank you for researching and implementing this!

davidhewitt commented:
Nit fixed, and rebased - expecting the CI to be green and fingers crossed this is good to merge!
